packages/squiggle-lang/src/rescript/GenericDist/GenericDist.res

```diff
@@ -0,0 +1,253 @@
+    operation: GenericDist_Types.Operation.arithmeticOperation,
+    t1: t,
+    t2: t,
+  ): option<result<SymbolicDistTypes.symbolicDist, string>> =>
+    switch (operation, t1, t2) {
+    | (operation, #Symbolic(d1), #Symbolic(d2)) =>
+      switch SymbolicDist.T.tryAnalyticalSimplification(d1, d2, operation) {
+      | #AnalyticalSolution(symbolicDist) => Some(Ok(symbolicDist))
+      | #Error(er) => Some(Error(er))
+      | #NoSolution => None
+      }
+    | _ => None
+    }
+
+  let runConvolution = (
+    toPointSet: toPointSetFn,
+    operation: GenericDist_Types.Operation.arithmeticOperation,
+    t1: t,
+    t2: t,
+  ) =>
+    E.R.merge(toPointSet(t1), toPointSet(t2))->E.R2.fmap(((a, b)) =>
+      PointSetDist.combineAlgebraically(operation, a, b)
+    )
+
+  let runMonteCarlo = (
+    toSampleSet: toSampleSetFn,
+    operation: GenericDist_Types.Operation.arithmeticOperation,
+    t1: t,
+    t2: t,
+  ) => {
+    let operation = Operation.Algebraic.toFn(operation)
+    E.R.merge(toSampleSet(t1), toSampleSet(t2))->E.R2.fmap(((a, b)) => {
+      Belt.Array.zip(a, b)->E.A2.fmap(((a, b)) => operation(a, b))
+    })
+  }
+
+  //I'm (Ozzie) really just guessing here, very little idea what's best
+  let expectedConvolutionCost: t => int = x =>
+    switch x {
+    | #Symbolic(#Float(_)) => 1
+    | #Symbolic(_) => 1000
+    | #PointSet(Discrete(m)) => m.xyShape->XYShape.T.length
+    | #PointSet(Mixed(_)) => 1000
+    | #PointSet(Continuous(_)) => 1000
+    | _ => 1000
+    }
+
+  let chooseConvolutionOrMonteCarlo = (t2: t, t1: t) =>
+    expectedConvolutionCost(t1) * expectedConvolutionCost(t2) > 10000
+      ? #CalculateWithMonteCarlo
+      : #CalculateWithConvolution
+
+  let run = (
+    t1: t,
+    toPointSet: toPointSetFn,
+    toSampleSet: toSampleSetFn,
+    algebraicOp,
+    t2: t,
+  ): result<t, error> => {
+    switch tryAnalyticalSimplification(algebraicOp, t1, t2) {
+    | Some(Ok(symbolicDist)) => Ok(#Symbolic(symbolicDist))
+    | Some(Error(e)) => Error(Other(e))
+    | None =>
+      switch chooseConvolutionOrMonteCarlo(t1, t2) {
+      | #CalculateWithMonteCarlo =>
+        runMonteCarlo(toSampleSet, algebraicOp, t1, t2)->E.R2.fmap(r => #SampleSet(r))
+      | #CalculateWithConvolution =>
+        runConvolution(toPointSet, algebraicOp, t1, t2)->E.R2.fmap(r => #PointSet(r))
+      }
+    }
+  }
+}
+
+let algebraicCombination = AlgebraicCombination.run
+
+//TODO: Add faster pointwiseCombine fn
+let pointwiseCombination = (t1: t, toPointSet: toPointSetFn, operation, t2: t): result<
+  t,
+  error,
+> => {
+  E.R.merge(toPointSet(t1), toPointSet(t2))
+  ->E.R2.fmap(((t1, t2)) =>
+    PointSetDist.combinePointwise(GenericDist_Types.Operation.arithmeticToFn(operation), t1, t2)
+  )
+  ->E.R2.fmap(r => #PointSet(r))
+}
+
+let pointwiseCombinationFloat = (
+  t: t,
+  toPointSet: toPointSetFn,
+  operation: GenericDist_Types.Operation.arithmeticOperation,
+  f: float,
+): result<t, error> => {
+  let m = switch operation {
```

Comment:
I don't have any suggestions for names other than `m`.

Reply:
So, stay with m?

Reply:
sure